### PR TITLE
Update SSO mapping providers documentation about unique IDs

### PR DESCRIPTION
**Pull request by clokep**
Tries to clarify some info based on a conversation with @anoadragon453 this morning.

### CODE REVIEW

**Comment by clokep** [comment]
I feel like we could use an introduction to the SSO modules, but I don't know where to put that...

File: docs/sso_mapping_providers.md [hunk 1]
```diff
@@ -67,8 +67,8 @@ A custom mapping provider must specify the following methods:
     - Arguments:
       - `userinfo` - A `authlib.oidc.core.claims.UserInfo` object to extract user
                      information from.
-    - This method must return a string, which is the unique identifier for the
-      user. Commonly the ``sub`` claim of the response.
+    - This method must return a string, which is the unique, immutable identifier
+      for the  user. Commonly the ``sub`` claim of the response.
```
Comment (anoadragon453) [inline docs/sso_mapping_providers.md:71]:
```suggestion
      for the user. Commonly the `sub` claim of the response.
```

File: docs/sso_mapping_providers.md [hunk 2]
```diff
@@ -153,8 +155,8 @@ A custom mapping provider must specify the following methods:
                           information from.
       - `client_redirect_url` - A string, the URL that the client will be
                                 redirected to.
-    - This method must return a string, which is the unique identifier for the
-      user. Commonly the ``uid`` claim of the response.
+    - This method must return a string, which is the unique, immutable identifier
+      for the user. Commonly the ``uid`` claim of the response.
```
Comment (anoadragon453) [inline docs/sso_mapping_providers.md:159]:
```suggestion
      for the user. Commonly the `uid` claim of the response.
```

File: docs/sso_mapping_providers.md [hunk 3]
```diff
@@ -172,8 +174,10 @@ A custom mapping provider must specify the following methods:
                                 redirected to.
     - This method must return a dictionary, which will then be used by Synapse
       to build a new user. The following keys are allowed:
-       * `mxid_localpart` - The mxid localpart of the new user.  If this is
-         `None`, the user is prompted to pick their own username.
+       * `mxid_localpart` - A string, the mxid localpart of the new user. If this is
+         `None`, the user is prompted to pick their own username. This is only used
+        during a user's first login. Once a localpart has been associated with a
+        remote user ID (see `get_remote_user_id`) it cannot be updated.
```
Comment (anoadragon453) [inline docs/sso_mapping_providers.md:180]:
```suggestion
       * `mxid_localpart` - A string, the mxid localpart of the new user. If this is
         `None`, the user is prompted to pick their own username. This is only used
         during a user's first login. Once a localpart has been associated with a
         remote user ID (see `get_remote_user_id`) it cannot be updated.
```

**Review by anoadragon453 — APPROVED** [review]
Definitely a footgun if people aren't aware about the immutability - so great to harp on about it, thanks!

I've also totally not taken the opportunity to do some final RST->MD here :stuck_out_tongue:

**Pull request merged**